chapter.md

```diff
@@ -0,0 +1,781 @@
+actually makes its changes easier to manage.
+
+## The Problem
+
+Users can interact with programs in a bewildering variety of ways,
+but the most popular on the web remains the HTML form.
+In a classic 1990s-style application,
+a user is presented with a form,
+fills in the fields,
+and clicks "Submit".
+This sends an HTTP POST request to a server
+with the information the user entered as its payload.
+The server validates that information,
+updates a database,
+and returns an HTML page that may or may not contain another form.
+
+As browsers have become more powerful,
+and as interaction has moved away from the desktop to mobile devices,
+developers have moved more and more of this processing cycle to the client.
+The first step was to write a little JavaScript
+to sanity-check the data the user entered before POSTing it
+so that users wouldn't waste time waiting for the server to respond,
+only to have to fix simple typos in the form.
+
+Programmers quickly began using JavaScript to make form interactions more fluid
+by showing and hiding sections based on what users had entered so far,
+and by fetching bits of data to do things like auto-completion while users were still typing.
+All of this made for a better user experience,
+but at the price of increasing client-side complexity.
+A modern single-page application (SPA) may have to keep track of dozens of pieces of information,
+any of which may need to be updated based on the user's actions
+*and* kept in sync with permanent storage on the server.
+
+As if that wasn't complicated enough,
+many of these interactions have to be done asynchronously
+in order to keep the user's FQ (frustration quotient) down.
+If,
+for example,
+a web page locks up for half a second every time the user types a single character
+because it's fetching possible auto-completes from the server,
+the user will quickly take her business somewhere else.
+
+Our problem is therefore this:
+
+> How should we manage client-side state in an asynchronous web application that uses forms?
+
+## Enter Redux
+
+Luckily for us,
+people who use pure functional programming languages
+have been thinking about these issues for more than thirty years,
+and have developed some design patterns that we can use
+in conventional languages like JavaScript.
+A *pure functional* language is one in which data cannot be changed,
+or *mutated*,
+in place:
+once a value is defined,
+it is *immutable*.
+Rather than changing its state,
+a program written in a pure functional language creates an entirely new state
+on which to operate.
+
+This may seem wasteful:
+after all,
+if I want to paint one wall of a room,
+I don't have to build an entirely new room
+that is identical to the original except for the color of the wall in question.
+However,
+making state immutable has a lot of advantages when we are dealing with concurrency.
+In particular,
+systems are a lot easier to reason about and test if they can't change under our feet.
```

Comment:
I think it might be worthwhile to throw in a brief example of mutable vs. immutable, perhaps using array splicing as an example